driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -734,6 +744,45 @@ public void channelClosed(ChannelHandlerContext ctx, ChannelStateEvent e) {
         }
     }
 
+    private static final ResponseCallback HEARTBEAT_CALLBACK = new ResponseCallback() {
+
+        @Override
+        public Message.Request request() {
+            return new Requests.Options();
+        }
+
+        @Override
+        public int retryCount() {
+            return 0; // no retries here
+        }
+
+        @Override
+        public void onSet(Connection connection, Message.Response response, long latency, int retryCount) {
+            switch (response.type) {
+                case SUPPORTED:
+                    logger.debug("{} heartbeat query succeeded", connection);
+                    break;
+                default:
+                    fail(connection, new ConnectionException(connection.address, "Unexpected heartbeat response: " + response));
+            }
+        }
+
+        @Override
+        public void onException(Connection connection, Exception exception, long latency, int retryCount) {
+            // Nothing to do: the connection is already defunct if we arrive here
+        }
+
+        @Override
+        public boolean onTimeout(Connection connection, long latency, int retryCount) {
+            fail(connection, new ConnectionException(connection.address, "Heartbeat query timed out"));
```

Comment:
This doesn't seem to be a problem as the connection is ultimately closed when force is called on the result of pool.closeAsync() in notifyOwnerWhenDefunct(), but I noticed that fail() here willl trigger the HostConnectionPool to call discardAvailableConnections() which calls tryTerminate on each connection in the pool.  tryTerminate doesn't do anything because we are in the middle of processing onTimeout, which means there is still an active ResponseHandler for this timeout in dispatcher.pending:

    DEBUG [2015-01-15 22:34:55,676] [Hashed wheel timer #1] Not terminating Connection[/127.0.0.2:9042-4, inFlight=0, closed=true]: there are still pending requests

Do we need to do any special handling here?  My first thought is no, but thought I'd bring it up.